### PR TITLE
Miscellaneous updates to training apis

### PR DESCRIPTION
**Pull request by baijumeswani**
This pull request updates the training apis by

- Renaming `ResetGrad` to `LazyResetGrad`
- Updating `RegisterLinearLRScheduler` documentation
- Adding `SetSeed` api for reproducibility.
- Adding apis to access training session input names and count

### CODE REVIEW

**Review by pengwa — APPROVED** [review]

**Comment by pengwa** [comment]
What's the motivation you make the 4th change?

**Comment by baijumeswani** [comment]
We have the apis to query the user graph output names already. We received a request to make apis available for querying user graph input names for debugging purposes. I think it can be useful information for the users.

File: orttraining/orttraining/training_api/include/module.h [hunk 1]
```diff
@@ -111,6 +111,18 @@ struct Module {
   Status ExportModelForInferencing(const std::string& inference_model_path,
                                    gsl::span<const std::string> graph_output_names) const;
 
+  // Returns the input count for training graph
```
Comment (askhade) [inline orttraining/orttraining/training_api/include/module.h:114]: nit: user input count. same comment applies for GetEvalModelInputCount api

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h [hunk 2]
```diff
@@ -11,310 +11,360 @@ ORT_RUNTIME_CLASS(CheckpointState);  /// Type that holds the training states for
...
+   * \param[in] sess The training session which has working knowledge of the optimizer model.
+   * \param[in] run_options Run options for this eval step.
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   */
   ORT_API2_STATUS(OptimizerStep, _Inout_ OrtTrainingSession* sess,
                   _In_opt_ const OrtRunOptions* run_options);
 
   /** \brief Registers the use of the Linear learning rate scheduler for the training session.
```
Comment (askhade) [inline orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:201]: nit: Registers Linear learning rate scheduler for the training session.

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h [hunk 3]
```diff
@@ -11,310 +11,360 @@ ORT_RUNTIME_CLASS(CheckpointState);  /// Type that holds the training states for
...
-  * \param[in] warmup_step_count Warmup steps for LR warmup.
-  * \param[in] total_step_count Total step count.
-  * \param[in] initial_lr The initial learning rate to be used by the training session.
-  *
-  * \snippet{doc} snippets.dox OrtStatus Return Value
-  *
-  */
+   *
+   * Register a Linear learning rate scheduler with the given learning rate scheduler parameters.
+   * Users must specify the initial learning rate that should be used with this learning rate scheduler
```
Comment (askhade) [inline orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:204]: Internally initial_lr is treated as optional:
Status TrainingSession::RegisterScheduler(const std::function<std::unique_ptr<LRSchedulerBase>std::shared_ptr<Optimizer>)>& get_scheduler,                            **std::optional<float>** initial_lr);

We should change this too or another option is to change the C API to accept float* to keep initial_lr as optional.

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h [hunk 4]
```diff
@@ -11,310 +11,360 @@ ORT_RUNTIME_CLASS(CheckpointState);  /// Type that holds the training states for
...
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   */
   ORT_API2_STATUS(ExportModelForInferencing, _Inout_ OrtTrainingSession* sess,
                   _In_ const ORTCHAR_T* inference_model_path, size_t graph_outputs_len,
                   _In_reads_(graph_outputs_len) const char* const* graph_output_names);
+
+  /** \brief Sets the seed used for random number generation in Onnxruntime.
+   *
+   * Use this to get reproducible results.
```
Comment (askhade) [inline orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:322]: nit: reproducible ->deterministic?

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h [hunk 5]
```diff
@@ -11,310 +11,360 @@ ORT_RUNTIME_CLASS(CheckpointState);  /// Type that holds the training states for
...
+   * \param[in] seed The seed to be set.
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   */
+  ORT_API2_STATUS(SetSeed, _In_ const int64_t seed);
+
+  /** \brief Retrieves the number of user inputs in the training model.
+   *
+   * This function returns the number of inputs of the training model.
```
Comment (askhade) [inline orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:333]: nit: 331 and 333 are duplicates

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h [hunk 6]
```diff
@@ -11,310 +11,360 @@ ORT_RUNTIME_CLASS(CheckpointState);  /// Type that holds the training states for
...
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   */
+  ORT_API2_STATUS(TrainingSessionGetTrainingModelInputCount, _In_ const OrtTrainingSession* sess, _Out_ size_t* out);
+
+  /** \brief Retrieves the number of user inputs in the eval model.
+   *
+   * This function returns the number of inputs of the eval model.
+   *
+   * \param[in] sess The training session which has working knowledge of the eval model.
```
Comment (askhade) [inline orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:347]: Shall we change "The training session which has working knowledge of the eval model." to "The training session which 
 owns the eval model." ?

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h [hunk 7]
```diff
@@ -11,310 +11,360 @@ ORT_RUNTIME_CLASS(CheckpointState);  /// Type that holds the training states for
...
+  /** \brief Sets the seed used for random number generation in Onnxruntime.
+   *
+   * Use this to get reproducible results.
+   *
+   * \param[in] seed The seed to be set.
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value
+   *
+   */
+  ORT_API2_STATUS(SetSeed, _In_ const int64_t seed);
```
Comment (askhade) [inline orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:329]: can you add a test for this?

File: orttraining/orttraining/python/training/api/module.py [hunk 8]
```diff
@@ -80,11 +80,11 @@ def eval(self):
         """
         return self.train(False)
 
-    def reset_grad(self):
+    def lazy_reset_grad(self):
         """
-        Resets the gradient of the parameters.
+        Lazily resets the gradient of the parameters.
```
Comment (askhade) [inline orttraining/orttraining/python/training/api/module.py:85]: can you add a little more explanation for what does lazy mean here... similar to C API comments.

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h (hunk 7 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:329]: for the api or for the functionality? The functionality tests are here: https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/test/framework/random_test.cc#L16

**Review by askhade — DISMISSED** [review]
LGTM

**Review by askhade — APPROVED** [review]
